tests/std/tests/VSO_0157762_feature_test_macros/test.cpp

```diff
@@ -5,465 +5,431 @@
-#define CXX17_MODE 1
-#define CXX20_MODE 0
-#else
-#define CXX17_MODE 0
-#define CXX20_MODE 0
+#error Expected __has_cpp_attribute to be defined.
 #endif
+#ifndef __has_cpp_attribute
+#error Expected __has_cpp_attribute to be defined.
 #endif
 
+#if __has_cpp_attribute(carries_dependency) != 200809L
+#error Expected has_cpp_attribute(carries_dependency) to equal 200809L.
+#endif
 
-// Always defined to varying values (C++14 versus C++17-and-newer mode).
+#if __has_cpp_attribute(deprecated) != 201309L
+#error Expected has_cpp_attribute(deprecated) to equal 201309L.
+#endif
 
-#ifndef __cpp_constexpr
-#error BOOM
-#elif CXX17_MODE
-#if __cpp_constexpr != 201603L
-#error BOOM
-#else
-STATIC_ASSERT(__cpp_constexpr == 201603L);
+#if __has_cpp_attribute(noreturn) != 200809L
+#error Expected has_cpp_attribute(noreturn) to equal 200809L.
+#endif
+
+#if defined(__clang__) || defined(__EDG__) // clang and EDG don't yet implement P1771R1
+#if __has_cpp_attribute(nodiscard) != 201603L
+#error Expected has_cpp_attribute(nodiscard) to equal 201603L.
 #endif
 #else
-#if __cpp_constexpr != 201304L
-#error BOOM
-#else
-STATIC_ASSERT(__cpp_constexpr == 201304L);
+#if __has_cpp_attribute(nodiscard) != 201907L
+#error Expected has_cpp_attribute(nodiscard) to equal 201907L.
 #endif
 #endif
 
-#ifndef __cpp_inheriting_constructors
-#error BOOM
-#elif (CXX17_MODE || defined(__clang__)) && !defined(__EDG__) // Clang implemented this C++17 feature unconditionally.
-                                                              // TRANSITION, VSO-610203
-#if __cpp_inheriting_constructors != 201511L
-#error BOOM
-#else
-STATIC_ASSERT(__cpp_inheriting_constructors == 201511L);
+#if _HAS_CXX17 || defined(__clang__) || defined(__EDG__) // Clang and EDG provide this in C++14 mode.
+#if __has_cpp_attribute(fallthrough) != 201603L
+#error Expected has_cpp_attribute(fallthrough) to equal 201603L.
 #endif
 #else
-#if __cpp_inheriting_constructors != 200802L
-#error BOOM
-#else
-STATIC_ASSERT(__cpp_inheriting_constructors == 200802L);
+#if __has_cpp_attribute(fallthrough) != 0
+#error Expected has_cpp_attribute(fallthrough) to equal 0.
 #endif
 #endif
 
-#ifndef __cpp_static_assert
-#error BOOM
-#elif CXX17_MODE
-#if __cpp_static_assert != 201411L
-#error BOOM
-#else
-STATIC_ASSERT(__cpp_static_assert == 201411L);
+#if _HAS_CXX17 || defined(__clang__) || defined(__EDG__) // Clang and EDG provide this in C++14 mode.
+#if __has_cpp_attribute(maybe_unused) != 201603L
+#error Expected has_cpp_attribute(maybe_unused) to equal 201603L.
 #endif
 #else
-#if __cpp_static_assert != 200410L
-#error BOOM
-#else
-STATIC_ASSERT(__cpp_static_assert == 200410L);
+#if __has_cpp_attribute(maybe_unused) != 0
+#error Expected has_cpp_attribute(maybe_unused) to equal 0.
 #endif
 #endif
 
+#if __has_cpp_attribute(noreturn) != 200809L
+#error Expected has_cpp_attribute(noreturn) to equal 200809L.
+#endif
 
-// Always defined to specific values.
+
+// COMPILER FEATURE-TEST MACROS

```

Comment:
I think I'd prefer "core" or "core language" to "compiler" here to reflect which part of the Standard defines these macros.

Reply:
Yeah I had "language" before realising I deleted the existing comment marking the section. Going with "core language"